src/test/java/ArraySortedBagTest.java

```diff
@@ -1,355 +1,381 @@
-        bag.add(99);
-        bag.add(99);
-        bag.add(99);
-        bag.remove(99);
-        assertEquals(2, bag.count(99));
-    }
+        @Test
+        void isEmpty_empty_returnsTrue() {
+            assertTrue(classUnderTest.isEmpty());
+        }
 
-    @Test
-    void removeFromSortedBagUpdatesSize() {
-        SortedBag<Integer> bag = new ArraySortedBag<>();
-        bag.add(99);
-        bag.add(101);
-        bag.remove(101);
-        assertEquals(1, bag.size());
-    }
+        @Test
+        void size_empty_returnsZero() {
+            assertEquals(0, classUnderTest.size());
+        }
 
-    @Test
-    void removeFromSortedBagWithSingleElementIsEmpty() {
-        SortedBag<Integer> bag = new ArraySortedBag<>();
-        bag.add(99);
-        bag.remove(99);
-        assertTrue(bag.isEmpty());
-    }
+        @Test
+        void iterator_empty_noNext() {
+            assertFalse(classUnderTest.iterator().hasNext());
+        }
 
-    @Test
-    void removeFromEmptySortedBagThrowsException() {
-        SortedBag<Integer> bag = new ArraySortedBag<>();
-        assertThrows(NoSuchElementException.class, () -> bag.remove(55));
-    }
+        @Test
+        void iterator_empty_nextThrowsNoSuchElementException() {
+            assertThrows(NoSuchElementException.class, () -> classUnderTest.iterator().next());
+        }
 
-    @Test
-    void removeNonExistingElementFromSortedBagThrowsException() {
-        SortedBag<Integer> bag = new ArraySortedBag<>();
-        bag.add(99);
-        bag.add(101);
-        assertThrows(NoSuchElementException.class, () -> bag.remove(55));
-    }
+        @Test
+        void toString_empty_returnsEmptyString() {
+            assertEquals("", classUnderTest.toString());
+        }
 
-    @Test
-    void iteratorHasNextOnEmptySortedBagReturnsFalse() {
-        SortedBag<Integer> bag = new ArraySortedBag<>();
-        Iterator<Integer> it = bag.iterator();
-        assertFalse(it.hasNext());
+        @Nested
+        class WhenSingleton {
+
+            static Stream<Arguments> elementsToAdd() {
+                return Stream.of(Arguments.of(0, List.of(0, 10)),
+                        Arguments.of(10, List.of(10, 10)),
+                        Arguments.of(20, List.of(10, 20)));
+            }
+
+            @BeforeEach
+            void addSingleton() {
+                classUnderTest.add(10);
+            }
+
+            @Test
+            void add_successfulAdd_returnsTrue() {
+                assertTrue(classUnderTest.add(0));
+            }
+
+            @ParameterizedTest
+            @MethodSource("elementsToAdd")
+            void add_singleton_correctLocation(int element, List<Integer> expected) {
+                List<Integer> list = new ArrayList<>();
+                classUnderTest.add(element);
+                while (!classUnderTest.isEmpty()) {
+                    list.add(classUnderTest.removeFirst());
+                }
+                assertEquals(expected, list);
+            }
```

Comment:
Although the adds in the setup code (`@BeforeEach`) could kinda' be done in such a way to test this, these explicit tests feel a little nicer since they allows for checking various situations (before, equal, middleish, end).

Reply:
I'm for it. Complex insertion, it's probably best to be explicit.